### ISSUE TITLE
Training an Equinox model that contains a PyTorch module using torchax - user define filter in eqx.filter_value_and_grad maybe?

### ISSUE THREAD

**Issue by Artur-Galstyan** [patrick-kidger/equinox#1137]
Ok, so this is going to be a bit more lengthy, sorry in advance!

Recently, I came across [torchax](https://github.com/google/torchax) and was happy to see that it might make any JAX-reimplementations of models as well as my [statedict2pytree](https://github.com/Artur-Galstyan/statedict2pytree) library obsolete.

The goal is to use torchax to train a JAX model, that contains some PyTorch modules. 

I managed to train the JAX part, but the PyTorch part remains frozen. This is what torchax advertises: using PyTorch models as feature extractors and this part works fine! (Although I could have just used `dlpack`, not 100% sure what the real benefit was?)

But I want to go one step further and actually train the PyTorch model that's inside my JAX model. In their README they allude to this but I couldn't get this to work.

This is the "MWE":

```python
from typing import Any

import equinox as eqx
import jax
import jax.numpy as jnp
import optax
import torch
import torchax
from torchax.interop import JittableModule
from torchvision import datasets, transforms

torchax.enable_globally()

# irrelevant MNIST data stuff
transform = transforms.Compose(
    [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
)
train_dataset = datasets.MNIST(
    root="./data", train=True, download=True, transform=transform
)
train_loader = torch.utils.data.DataLoader(
    train_dataset, batch_size=32, shuffle=True, drop_last=True
)


class TorchPart(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        self.linear = torch.nn.Linear(in_features, out_features)

    def forward(self, x):
        return self.linear(x)


class HybridModel(eqx.Module):
    jax_linear1: eqx.nn.Linear
    torch_part: JittableModule
    jax_linear2: eqx.nn.Linear

    def __init__(self, key):
        self.jax_linear1 = eqx.nn.Linear(784, 128, key=key)
        torch_model = TorchPart(128, 128).to("jax")
        self.torch_part = JittableModule(torch_model)
        self.jax_linear2 = eqx.nn.Linear(128, 10, key=key)

    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        x = x.reshape(-1, 784)
        x = eqx.filter_vmap(self.jax_linear1)(x)
        x = x.reshape(-1, 128)
        x = self.torch_part(x).jax()
        x = jax.nn.relu(x)
        x = eqx.filter_vmap(self.jax_linear2)(x)
        return x


def loss_fn(model: HybridModel, x: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:
    logits = model(x)
    return optax.softmax_cross_entropy_with_integer_labels(
        logits=logits, labels=y
    ).mean()


def torchax_filter(element: Any):
    if eqx.is_array(element):
        return True
    if isinstance(element, torch.Tensor):
        return True
    return False


@eqx.filter_jit
def train_step(
    model: HybridModel,
    opt_state: Any,
    optimizer: optax.GradientTransformation,
    x: jnp.ndarray,
    y: jnp.ndarray,
) -> tuple[HybridModel, Any, jnp.ndarray]:
    (loss_val, grads) = eqx.filter_value_and_grad(loss_fn)(model, x, y)

    updates, opt_state = optimizer.update(grads, opt_state, model)
    model = eqx.apply_updates(model, updates)
    return model, opt_state, loss_val


main_key = jax.random.key(42)
model = HybridModel(main_key)

optimizer = optax.adam(1e-3)
opt_state = optimizer.init(eqx.filter(model, torchax_filter))

linear_before_training = model.torch_part.params["linear.weight"].jax()
print("before training", linear_before_training)


for epoch in range(1):
    for batch_idx, (x_torch, y_torch) in enumerate(train_loader):
        x_jax = jnp.array(x_torch.numpy())
        y_jax = jnp.array(y_torch.numpy())

        model, opt_state, loss = train_step(model, opt_state, optimizer, x_jax, y_jax)

        if batch_idx % 100 == 0:
            print(f"Batch {batch_idx}, Loss: {loss.item()}")

        if batch_idx >= 500:
            break


linear_after_training = model.torch_part.params["linear.weight"].jax()
print("after training", linear_after_training)
```

A snippet from the output shows e.g. this 

```
before training [[-0.05707549 -0.08610713 -0.04081073 ...  0.07019104 -0.05778148

...

after training [[-0.05707549 -0.08610713 -0.04081073 ...  0.07019104 -0.05778148
```

Which are the same numbers. But the loss is going down, so the JAX part is actually training. 

My initial guess as to why it's skipping the PyTorch weights is because `eqx.filter_value_and_grad` filters using the following:

```python
diff_x, nondiff_x = partition(x, is_inexact_array)
```

and `is_inexact_array` filters like this

```python
def is_inexact_array(element: Any) -> bool:
    """Returns `True` if `element` is an inexact (i.e. floating or complex) JAX/NumPy
    array.
    """
    if isinstance(element, (np.ndarray, np.generic)):
        return bool(np.issubdtype(element.dtype, np.inexact))
    elif isinstance(element, jax.Array):
        return jnp.issubdtype(element.dtype, jnp.inexact)
    else:
        return False
```

Therefore, the gradients actually look like this:

```
HybridModel(
  jax_linear1=Linear(
    weight=f32[128,784],
    bias=f32[128],
    in_features=784,
    out_features=128,
    use_bias=True
  ),
  torch_part=None, <<< oh no!
  jax_linear2=Linear(
    weight=f32[10,128],
    bias=f32[10],
    in_features=128,
    out_features=10,
    use_bias=True
  )
)
```

I thought if I added a customer filter to the `filter_value_and_grad` function it would compute the gradients:

```python
        def torchax_filter(element: Any):
            import torch
            if is_inexact_array(element):
                return True
            if isinstance(element, torch.Tensor):
                return True
            return False

        diff_x, nondiff_x = partition(x, torchax_filter)
        return fun_value_and_grad(diff_x, nondiff_x, *args, **kwargs)
```

But to no avail. I feel like I'm sooo close, I just need JAX to compute the gradients for me and then all the current problems are solved (state management is a whole different beast, but for now I just want the MWE to work). 

Has anyone got torchax + equinox + PyTorch to work yet?

**Comment by johannahaffner** [comment]
Hi! 

This is a large example, and I don't have time right now to run it myself.

That said, I would first try to get a gradient of the Torch submodule in JAX, without considering the complexity of all the other modules. Defining a custom filter might not be enough here, JAX requires the correct primitives to compute gradients. You could try defining a custom JVP for your submodule (or a VJP if that is more appropriate), and then see if you can get JAX gradients for the Torch submodule this way. 

As soon as you can get a JAX gradient for this submodule, you should be able to get a gradient using `eqx.filter_grad` (or another filtered transformation), perhaps with a custom filter to make sure that the Torch module is not treated as static. And once this works, adding back the other modules should not be a problem. 

Fingers crossed!

**Comment by Artur-Galstyan** [comment]
@johannahaffner Turns out, yes, we can use torchax and Equinox and PyTorch together!

I made a gist, which trains the ResNet (from PyTorch) inside an Equinox model, which can be found [here](https://gist.github.com/Artur-Galstyan/e3c3f5a809b8178ded4809649f3aa861). This requires version 0.0.9 in torchax, otherwise you will run into this [issue](https://github.com/google/torchax/issues/32). I want to try out other, more complex models (RFDiffusion, SAM, ESMFold, LLAMA and others) and see how it goes, but it looks promising.

**Comment by johannahaffner** [comment]
Very nice! Cool gist, this will help a lot of people who come across this issue looking for pointers. Thanks for sharing it!

**Comment by adam-hartshorne** [comment]
Does Torchax have an advantage over using Torch2Jax (https://github.com/rdyro/torch2jax) ?

**Comment by Artur-Galstyan** [comment]
I haven't used that one yet, so I can't really comment on that:/
